preprocessing/structs.ts

```diff
@@ -1,15 +1,82 @@
     
    **Let the devs know about any error, typo etc, by contacting 	devs@mangrove.exchange**
+
+
+  There is one Mangrove contract that manages all tradeable offerLists. This reduces deployment costs for new offerLists and lets market makers have all their provision for all offerLists in the same place.
+
+   The interaction map between the different actors is as follows:
+   <img src="./contactMap.png" width="190%"></img>
+
+   The sequence diagram of a market order is as follows:
+   <img src="./sequenceChart.png" width="190%"></img>
+
+
+   ## `tickSpacing`
+
+   The granularity of available price points in an offer list is controlled by the `tickSpacing` parameter. With a high `tickSpacing`, fewer price points are available, but the gas cost of market orders is lower since a smaller part of the tick tree has to be explored.
+
+   The available prices in an offer list are `1.0001^i` for all `MIN_TICK <= i <= MAX_TICK` such that `i % tickSpacing = 0`.
+
+   ## Tree storage
+
+   The exposed structure of offers is a list, but offers are actually stored in a tree we call a "tick tree". 
+
+   Thanks to the tree structure, offer insertion takes constant time (the height of the tree is fixed).
+
+   ### Bins
+
+   <img src="./bin.png" width="80%"></img>
+
+   Below the bottom of the tree are _bins_. A bin is a doubly-linked list of offers. All offers in a bin have the same tick. During a market order, offers in a bin are executed in order, from the first to the last. Inserted offers are always appended at the end of a bin. 
+
+   Bins are laid in sequence. If all the offers in a bin have tick `t`, all the offers in the following bin have tick `t+tickSpacing`.
+
+   Bins are identified by their index in the bin sequence, from the first (`MIN_BIN`) to the last (`MAX_BIN`). The number of available bins is larger than the number of available ticks, so some bins will never be used.
+
+   ### Tree structure
+
+   The structure of the tick tree is as follows:
+   <img src="./tick_tree_structure.png" width="190%"></img>
+
+   At the bottom of the tree, leaves contain information about 4 bins: their first and last offer. Offer ids use 32 bits, so leaves use 256 bits.
+
+   When a market order runs, execution starts with the first offer of the smallest nonempty bin and continues from there.
+
+   Once all the offers in the smallest bin have been executed, the next nonempty bin is found. If the leaf of the current bin now only has empty bins, the node above the current leaf contains a bitfield with information about all its children: if the *i*th bit of the node is set, its *i*th child has at least one nonempty bin. Otherwise, its *i*th child only has empty bins.
+
+   To find the next nonempty bin, it may be necessary to keep going up the tree until the root is reached. At each level above, the bitfield rule applies similarly: the *i*th bit of a node is set iff its *i*th child has at least one set bit.
```

Comment:
> If the leaf of the current bin now only has empty bins, the node above the current leaf contains a bitfield with information about all its children

The node above always contains this bitfield :-)

Maybe something like this instead:

```suggestion
   Once all the offers in the smallest bin have been executed, the next non-empty bin is found. If the leaf of the current bin now only has empty bins, the tree must be searched for the next non-empty bin, starting at the node above the leaf:
   
   A non-leaf node contains a bit field with information about all its children: if the *i*th bit of the node is set, its *i*th child has at least one non-empty bin. Otherwise, its *i*th child only has empty bins.

   To find the next non-empty bin, it may be necessary to keep going up the tree until the root is reached. At each level above, the bit field rule applies similarly: the *i*th bit of a node is set iff its *i*th child has at least one set bit.
```